Feature/tracing

akoerner1 opened the pull request:
## Description
This PR introduces user and kernel-level tracing with the support of [ros2_tracing](https://github.com/ros2/ros2_tracing)
and [ROS 2 Observer](https://github.com/DLR-TS/ros2_observer).

All node tracing and inspection data will be provided in the `ROS_HOME/ros2_node_inspector/` and `ROS_HOME/ros2_tracer/` directories. In the ADORe CLI the ROS_HOME is set to `./.log/.ros` or `/tmp/adore/.log/.ros` 


### Key Features
- Nodes can be "inspected" with the `ros2-node-inspector` command (see: [https://github.com/DLR-TS/ros2_observer/blob/a75db3ca80de8ad9df6260faa3a83e9d357cfea0/README.md?plain=1#L132](https://github.com/DLR-TS/ros2_observer/blob/a75db3ca80de8ad9df6260faa3a83e9d357cfea0/README.md?plain=1#L132))
  Usage:
   ```
   make cli
   ros2-node-tracer
   ```
  - Node data including publishers, subscribers, interfaces, and datatypes are parsed and provided as JSON.
  - Node data is used to generate a connected graph as a JSON and a GraphViz DOT file
  Example output:
  ![image](https://github.com/user-attachments/assets/a57fa5ee-9ab4-4e21-a38c-338347195f28)

- User-level and kernel-level program tracing data: The [ros2_tracing](https://github.com/ros2/ros2_tracing) has been integrated into the ADORe CLI via [ROS 2 Observer](https://github.com/DLR-TS/ros2_observer) 
  - supports live and timed tracing sessions
  - auto-conversion of trace data to CTF format and JSON using `babeltrace2`
  Example trace data JSON excerpt:
```
[
    {
        "timestamp": "15:15:17.584761455",
        "delta": "+?.?????????",
        "hostname": "ADORe-CLI",
        "event": "ros2:rclcpp_executor_get_next_ready",
        "data": "{ cpu_id = 8 }, { procname = \"visualizer\", vpid = 2242303, vtid = 2242303 }, { }",
        "cpu_id": "8",
        "procname": "visualizer",
        "vpid": "2242303",
        "vtid": "2242303"
    },
    {
        "timestamp": "15:15:17.584765452",
        "delta": "+0.000003997",
        "hostname": "ADORe-CLI",
        "event": "ros2:rclcpp_executor_execute",
        "data": "{ cpu_id = 8 }, { procname = \"visualizer\", vpid = 2242303, vtid = 2242303 }, { handle = 0x5BF470BA3950 }",
        "cpu_id": "8",
        "procname": "visualizer",
        "vpid": "2242303",
        "vtid": "2242303",
        "handle": "0x5BF470BA3950"
    },
...
]
```
- Live trace data streaming via the "lttng-live" protocol

 Review the ROS 2 Observer README for more information on features:
[https://github.com/DLR-TS/ros2_observer/blob/main/README.md](https://github.com/DLR-TS/ros2_observer/blob/main/README.md)


### New `make trace` Target
Running `make trace` will now execute a tracing session via the ROS 2 Observer
Assuming a scenario is executing run the following command at the root of the ADORe project:
```
make trace
```

This will initiate a tracing session wherein user-level and kernel-level trace data will be collected for 
**5** seconds. After this data collection session the program `Trace Compass` will be automatically launched
for analysis of the tracing data:
![image](https://github.com/user-attachments/assets/2130555e-c91b-4a06-90e3-6f5949b2646e)

Alternatively, Lttng Scope can be used for trace analysis with the following:
```
source adore.env
cd vendor/ros2_observer/lttng_scope
make start
```
![image](https://github.com/user-attachments/assets/1e097772-450f-4838-a72a-56cb21b9b6bd)

**Note:** Lttng Scope does not support auto-opening a trace path. Once the program is launched the trace must be manually
opened with the path `/opt/lttng_scope/tracing_logs/ros2_tracer/trace.ctf`

### Live Tracing With ROS 2 Observer and `ros2_tracing`
With the help of the [ROS 2 Observer](https://github.com/DLR-TS/ros2_observer) lttng-live tracing sessions can be started.
1. Start a scenario:
```
make cli
cd adore_scenarios/simulation_scenarios
ros2 launch simulation_test.py
```
2. Start the live tracing session, in another shell run:
```
make cli
ros2-tracer --live -o
```
3. Use an observer to connect to the session. For example you can use the included distribution of `babeltrace2` provided by the ADORe CLI:
```
make cli
babeltrace2 --input-format=lttng-live net://localhost/host/ADORe-CLI/ros2_tracer
```

## Type of Change

Please delete options that are not relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (adds or updates documentation)
- [ ] Refactor (non-breaking change for code readability/structure)
- [ ] Other (please describe):

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] Tests pass locally with my changes.
- [x] I have updated the documentation where necessary.

**Test Configuration**:
- OS: Ubuntu 24.04
- Docker Version: Docker version 27.3.1, build ce12230

MikkelDLR committed:
Develop XODR support into master (#7)

libopendrive vendor, and XODR map support and release building

---------

Co-authored-by: Andrew Koerner <99262104+akoerner1@users.noreply.github.com>
Co-authored-by: Andrew Koerner <andrew.koerner@dlr.de>
Co-authored-by: markodlr <marko.mizdrak@dlr.de>

akoerner1 committed:
added ros2_observer and ros2_tracing

akoerner1 committed:
merged develop e61e69e

akoerner1 committed:
update .gitmodules

akoerner1 committed:
update ROS_HOME env var

markodlr commented:
Interesting looking tool.

It seems that the gitmodules file is incorrectly including the ros2_observer submodule twice.

akoerner commented:
- fix timestamp bug
- fix gitmodules

n-mat committed:
initial version of carla_bridge added

n-mat committed:
updated submodule

akoerner1 committed:
removed duplicate entry in the gitmodules for the ros2_observer, update the ros2_observer to incorperate a bugfix in the trace parser involving incorrect timestamps

akoerner1 committed:
Merge branch 'develop' into feature/tracing

n-mat committed:
Merge remote-tracking branch 'origin/develop' into carla-bridge

n-mat committed:
fix of includes

akoerner1 committed:
updated adore_cli + make_gadgets, fixed DOCKER_GID (#14)

## Description

- Updated tools/adore_cli/make_gadgets to improve the DOCKER_GID make
variable. If there are multiple entries in the /etc/group for docker the
previous implementation would return multiple group ids. This change
makes it more reliable if there are multiple entries in the /etc/group
file for the same docker group.

 - Also added .dockerignore to libOpenDRIVE to improve docker caching
- Update the `adore_cli` .zshrc file to include git hash info in the
PROMPT
Old Prompt:

![image](https://github.com/user-attachments/assets/3d18f6f2-11c4-4864-8197-a2bb554cb2e4)
New Prompt:

![image](https://github.com/user-attachments/assets/ba4e5788-569d-412a-ad80-f4dfce6798e8)

- Updated/changed `adore_cli` container name and image names

https://github.com/DLR-TS/adore_cli/blob/fcc2fd0eb510d77bff964f95a5869b8f19d63df2/adore_cli.mk#L27-L36
  

## Type of Change

Please delete options that are not relevant:

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- [ ] Documentation update (adds or updates documentation)
- [ ] Refactor (non-breaking change for code readability/structure)
- [ ] Other (please describe):

akoerner1 committed:
merged with origin/develop

akoerner1 committed:
Merge branch 'feature/tracing' of github.com:DLR-TS/adore into feature/tracing

akoerner1 committed:
removed syntax error from ros2_hello_world.cpp

akoerner1 committed:
added ros2_observer to the git modules

n-mat committed:
update of libraries submodule

n-mat committed:
updated mission_control submodule

n-mat committed:
updated carla_bridge submodule

n-mat committed:
added requirement for carla

akoerner1 committed:
updated Makefile

akoerner1 committed:
updated Makefile

akoerner1 committed:
updated Makefile

akoerner1 committed:
updated Makefile

akoerner1 committed:
updated Makefile

akoerner1 committed:
updated Makefile

n-mat committed:
CARLA bridge (#15)

## Description

This PR adds the carla_bridge

## Type of Change
- [X] New feature (non-breaking change which adds functionality)


## Checklist

- [X] I have performed a self-review of my own code.
- [X] I have commented my code, particularly in hard-to-understand
areas.
- [X] I have updated the documentation where necessary.


## Notes for Reviewers

As discussed, low level control needs to be fixed before merging.

jnboeg committed:
Fix: Landing page footer and refactoring of stylesheet (#18)

## Description

Please provide a brief summary of the changes you made and why they are
necessary.

Fixed issue of the landing page where the footer would not scale
correctly in different browsers and screen resolutions.
Refactoring of the css stylesheet, elements with identical
characteristics are now grouped together in css file.

## Type of Change

Please delete options that are not relevant:

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Refactor (non-breaking change for code readability/structure)

## Checklist

- [X] I have performed a self-review of my own code.
- [X] Tests pass locally with my changes.
- [X] I have updated the documentation where necessary.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide
instructions so others can reproduce.

- Tested in multiple Browsers
- Tested by publishing landing page for fork

**Test Configuration**:
- OS: Fedora 41
- Docker Version: 28.1.1

MikkelDLR committed:
checkpoing merger

MikkelDLR committed:
updated adore scenarios hash
